src/net_processing.cpp

```diff
@@ -2496,18 +2473,14 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
             // nodes)
             m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::SENDHEADERS));
         }
-        if (pfrom.GetCommonVersion() >= SHORT_IDS_BLOCKS_VERSION) {
-            // Tell our peer we are willing to provide version 1 or 2 cmpctblocks
+        if (pfrom.GetCommonVersion() >= SHORT_IDS_BLOCKS_VERSION &&
+            WITH_LOCK(cs_main, {return State(pfrom.GetId())->fHaveWitness;})) {
```

Comment:
> can we avoid locking the mutex

Very nice observation, although I think this is ok for now. We take cs_main when processing almost all messages in net_processing (although I agree with you that it'd be slightly nicer if we didn't have to). In the new code, we're only sending `sendcmpct` if the peer has told us that they support witness serialization. That's indicated by fHaveWitness or pfrom.nServices (pfrom.GetLocalServices() is something different - it's the services we offered to the peer).

fHaveWitness will move into Peer very soon as part of #19398, at which point it'll no longer be guarded by cs_main.

> (and order dependance on fHaveWitness being set)

`fHaveWitness` can only be set once in version handling, and this code in verack handling can only be hit after that.

Reply:
> Maybe we can rename SHORT_IDS_BLOCK_VERSION to SENDCMPCT_VERSION

I do have a branch somewhere that does that, but it's out of scope for this PR.

Reply:
> I do have a branch somewhere that does that, but it's out of scope for this PR.

Here you go: https://github.com/jnewbery/bitcoin/commit/82372335cc3cf8fa238678629dda6a2f465e6005. Feel free to PR that commit if you think it's worth it.

Reply:
Agree with @jonatack about taking from `(pfrom.nServices & NODE_WITNESS)` rather than `State(pfrom.GetId())->fHaveWitness;})` just because it's easier to not have to worry about where the latter comes from.

Understood that it's moot as it's changing soon anyway, so not a big deal.